src/hotspot/share/opto/loopopts.cpp

```diff
@@ -1512,23 +1515,32 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
               if (in != NULL && n_loop->is_member(get_loop(get_ctrl(in)))) {
                 const Type* in_t = _igvn.type(in);
                 if (in_t->isa_int()) {
-                  cast = new CastIINode(in, in_t, true);
+                  cast = new CastIINode(in, in_t, ConstraintCastNode::UnconditionalDependency);
                 } else if (in_t->isa_long()) {
-                  cast = new CastLLNode(in, in_t, true);
+                  cast = new CastLLNode(in, in_t, ConstraintCastNode::UnconditionalDependency);
                 } else if (in_t->isa_ptr()) {
-                  cast = new CastPPNode(in, in_t, true);
+                  cast = new CastPPNode(in, in_t, ConstraintCastNode::UnconditionalDependency);
                 } else if (in_t->isa_float()) {
-                  cast = new CastFFNode(in, in_t, true);
+                  cast = new CastFFNode(in, in_t, ConstraintCastNode::UnconditionalDependency);
                 } else if (in_t->isa_double()) {
-                  cast = new CastDDNode(in, in_t, true);
+                  cast = new CastDDNode(in, in_t, ConstraintCastNode::UnconditionalDependency);
                 } else if (in_t->isa_vect()) {
-                  cast = new CastVVNode(in, in_t, true);
+                  cast = new CastVVNode(in, in_t, ConstraintCastNode::UnconditionalDependency);
                 }
               }
               if (cast != NULL) {
                 cast->set_req(0, x_ctrl);
                 register_new_node(cast, x_ctrl);
                 x->replace_edge(in, cast);
+                if (x->is_AddP() && k == AddPNode::Base) {
+                  for (DUIterator_Fast imax, i = x->fast_outs(imax); i < imax; i++) {
+                    Node* u = x->fast_out(i);
+                    if (u->is_AddP()) {
```

Comment:
I think we need to check that originally both AddP point to the same base.
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/escape.cpp#L2502

Reply:
> I think we need to check that originally both AddP point to the same base.

Right. I pushed an updated change with a fix for that.